iroh-net/src/discovery/local_node_discovery.rs

```diff
@@ -0,0 +1,250 @@
+                                );
+                                continue;
+                            }
+                        };
+
+                        if discovered_node_id == node_id {
+                            continue;
+                        }
+
+                        if peer_info.is_expiry() {
+                            tracing::trace!(
+                                ?discovered_node_id,
+                                "removing node from LocalNodeDiscovery address book"
+                            );
+                            node_addrs.remove(&discovered_node_id);
+                            continue;
+                        }
+
+                        if let Some(sender) = senders.get(&discovered_node_id) {
+                            let item: DiscoveryItem = (&peer_info).into();
+                            tracing::trace!(?item, "sending DiscoveryItem");
+                            sender.send_async(Ok(item)).await.ok();
+                        }
+                        tracing::trace!(
+                            ?discovered_node_id,
+                            ?peer_info,
+                            "adding node to LocalNodeDiscovery address book"
+                        );
+                        node_addrs.insert(discovered_node_id, peer_info);
+                    }
+                    Message::SendAddrs((node_id, sender)) => {
+                        tracing::trace!(?node_id, "LocalNodeDiscovery Message::SendAddrs");
+                        if let Some(peer_info) = node_addrs.get(&node_id) {
+                            let item: DiscoveryItem = peer_info.into();
+                            tracing::debug!(?item, "sending DiscoveryItem");
+                            sender.send_async(Ok(item)).await.ok();
+                        }
+                        senders.insert(node_id, sender);
+                        let timeout_sender = task_sender.clone();
+                        timeouts.spawn(async move {
+                            tokio::time::sleep(DISCOVERY_DURATION).await;
+                            tracing::trace!(?node_id, "discovery timeout");
+                            timeout_sender
+                                .send_async(Message::Timeout(node_id))
+                                .await
+                                .ok();
+                        });
+                    }
+                    Message::Timeout(node_id) => {
+                        tracing::trace!(?node_id, "LocalNodeDiscovery Message::Timeout");
+                        senders.remove(&node_id);
+                    }
+                    Message::ChangeLocalAddrs(addrs) => {
+                        tracing::trace!(?addrs, "LocalNodeDiscovery Message::ChangeLocalAddrs");
+                        let callback_send = task_sender.clone();
+                        let g = guard.take();
+                        drop(g);
```

Comment:
One of the next things I wanted to do is to add an API for changing the locally advertised addresses without stopping the service; this has the advantage of not losing the swarm size estimate and I just realised that this current approach means that expiration isn’t signalled reliably.

In any case, it is good to see this example before working on that feature. Will let you know when I have implemented it.

Reply:
Oh yes this would be a great API change 🙌 thanks!